### PR TITLE
BN-833-quivr-rules-proof-proposition-and-shared

### PR DESCRIPTION
**Pull request by nandotorterolo**
## Purpose
quivr protobuf questions rules
## Approach
implement rules
## Testing
Compile and publish
## TODO
solve alls:  /**/ ask if we need len or max len rule**

## Tickets
 https://topl.atlassian.net/browse/BN-833

### CODE REVIEW

File: proto/quivr/models/shared.proto [hunk 1]
```diff
@@ -2,25 +2,29 @@ syntax = "proto3";
 
 package quivr.models;
 
+import "validate/validate.proto";
+import "scalapb/scalapb.proto";
+import "scalapb/validate.proto";
+
 message Data {
-    bytes value = 1;
+    bytes value = 1;  // ask if we need len or max len rule
```
Comment (scasplte2) [inline proto/quivr/models/shared.proto:10]: should match any size limit in `ExactMatch, LessThan, GreaterThan, and EqualTo` or we might want to chat about this as I recall seeing that there might be a conflict with the size for Transaction data.

**Review by scasplte2 — APPROVED** [review]
couple of responses but nothing too much

File: proto/quivr/models/proposition.proto [hunk 2]
```diff
@@ -49,37 +54,53 @@ message Proposition {
 
     message ExactMatch {
         string location = 1;
-        bytes compareTo = 2;
+        bytes compareTo = 2; // ask if we need len or max len rule
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply proto/quivr/models/proposition.proto:57]: I think a max length of `64` is a fair starting point.

File: proto/quivr/models/proposition.proto [hunk 3]
```diff
@@ -49,37 +54,53 @@ message Proposition {
...
     }
 
     message EqualTo {
         string location = 1;
-        Int128 compareTo = 2;
+        Int128 compareTo = 2 [(validate.rules).message.required = true];
     }
 
     message Threshold {
         repeated Proposition challenges = 1;
```
[the remark this answers is not in the document]
Comment (SeanCheatham) [reply proto/quivr/models/proposition.proto:76]: Perhaps we could require `challenges` to be non-empty (min-length of 1).  The result is a `Seq[Proposition]`.  A `Proposition` self-encodes Optionality by generating a `Proposition.Empty` as a valid value.  I'm not aware of a way to describe on protoc-gen-validate syntax the following: `challenges.forall(_.value != Proposition.Empty)`.  I think we'll need to do that one manually.

File: proto/quivr/models/shared.proto (hunk 1 above)
Comment (SeanCheatham) [reply proto/quivr/models/shared.proto:10]: I think that'll make things interesting.  I think if something like `Transaction.Data` can be used in an `EqualTo`, then it should just be the same limit across the board.

File: proto/quivr/models/shared.proto [hunk 4]
```diff
@@ -2,25 +2,29 @@ syntax = "proto3";
...
-        bytes root64 = 2;
+        bytes root32 = 1 [(validate.rules).bytes.len = 32];
+        bytes root64 = 2 [(validate.rules).bytes.len = 64];
     }
 }
 
 message Preimage {
-    bytes input = 1;
-    bytes salt = 2;
+    bytes input = 1; // ask if we need len or max len rule
```
Comment (SeanCheatham) [inline proto/quivr/models/shared.proto:26]: For safety, perhaps we should limit both of these fields to something _relatively high_ like 16kb

File: proto/quivr/models/shared.proto [hunk 5]
```diff
@@ -31,57 +35,73 @@ message Digest {
...
 message DigestVerification {
-    Digest digest = 1;
-    Preimage preimage = 2;
+    Digest digest = 1 [(validate.rules).message.required = true];
+    Preimage preimage = 2 [(validate.rules).message.required = true];
 }
 
 message VerificationKey {
-    bytes value = 1;
+    bytes value = 1 [(validate.rules).bytes.len = 32];
```
Comment (SeanCheatham) [inline proto/quivr/models/shared.proto:53]: @scasplte2 Should we strictly define the VK/SK lengths?  Or should we instead limit it to something higher and then list some common lengths in a comment?

File: proto/quivr/models/shared.proto [hunk 6]
```diff
@@ -31,57 +35,73 @@ message Digest {
...
-    Witness signature = 2;
-    Message message = 3;
+    VerificationKey verificationKey = 1 [(validate.rules).message.required = true];
+    Witness signature = 2 [(validate.rules).message.required = true];
+    Message message = 3 [(validate.rules).message.required = true];
 }
 
 message SignableBytes {
-    bytes value = 1;
+    bytes value = 1;  // ask help for len, max len rule
```
Comment (SeanCheatham) [inline proto/quivr/models/shared.proto:80]: I don't think this value will be transmitted, so I think we can leave it unbounded for now

File: proto/quivr/models/shared.proto [hunk 7]
```diff
@@ -31,57 +35,73 @@ message Digest {
...
 }
 
 message TxBind {
-    bytes value = 2;
+    bytes value = 2; //  ask help for len, max len rule, index should be 1?
```
Comment (SeanCheatham) [inline proto/quivr/models/shared.proto:84]: I'm unsure on this one

**Review by SeanCheatham — APPROVED** [review]

File: proto/quivr/models/proposition.proto [hunk 8]
```diff
@@ -49,37 +54,53 @@ message Proposition {
 
     message ExactMatch {
         string location = 1;
```
[the remark this answers is not in the document]
Comment (nandotorterolo) [reply proto/quivr/models/proposition.proto:56]: thanks for review it, string, is a required field, no rules is needed

File: proto/quivr/models/shared.proto (hunk 5 above)
Comment (scasplte2) [reply proto/quivr/models/shared.proto:53]: specific lengths would be safest but would expand quickly I imagine (but perhaps not?). I don't have the strongest preference so I'd say whichever you all judge is the best compromise on scalability / type safety.

File: proto/quivr/models/shared.proto (hunk 7 above)
Comment (scasplte2) [reply proto/quivr/models/shared.proto:84]: the bind length depends on the length of the specified hash algorithm that is chosen in Quivr. At the moment that is Blake2b256 I believe, so we could fix this to 32. Likely could get away with a similar approach as Sean suggested before with an upper bound if we want to be most flexible. I don't know the prevalence of hash lengths over 64 bytes personally.

**Pull request merged**